src/VisualStudio/Core/Def/Implementation/IntellisenseControls/IntellisenseTextBoxViewModelFactory.cs

```diff
@@ -0,0 +1,129 @@
+            var contextBuffer = _projectionBufferFactoryService.CreateProjectionBuffer(null,
+                new object[] { originalContextBuffer.CurrentSnapshot.CreateFullTrackingSpan(SpanTrackingMode.EdgeInclusive) }, ProjectionBufferOptions.None, contentType);
+
+            // Make projection readonly so we can't edit it by mistake.
+            using (var regionEdit = contextBuffer.CreateReadOnlyRegionEdit())
+            {
+                regionEdit.CreateReadOnlyRegion(new Span(0, contextBuffer.CurrentSnapshot.Length), SpanTrackingMode.EdgeInclusive, EdgeInsertionMode.Deny);
+                regionEdit.Apply();
+            }
+
+            // Put it all into a projection buffer
+            var projectionBuffer = _projectionBufferFactoryService.CreateProjectionBuffer(null,
+                createSpansMethod(contextBuffer.CurrentSnapshot),
+                ProjectionBufferOptions.None, contentType);
+
+            // Fork the solution using this new primary buffer for the document and all of its linked documents.
+            var forkedSolution = solution.WithDocumentText(document.Id, projectionBuffer.CurrentSnapshot.AsText(), PreservationMode.PreserveIdentity);
+            foreach (var link in document.GetLinkedDocumentIds())
+            {
+                forkedSolution = forkedSolution.WithDocumentText(link, projectionBuffer.CurrentSnapshot.AsText(), PreservationMode.PreserveIdentity);
+            }
+
+            // Put it into a new workspace, and open it and its related documents
+            // with the projection buffer as the text.
+            var workspace = new IntellisenseTextBoxWorkspace(forkedSolution, document.Project);
+            workspace.OpenDocument(workspace.ChangeSignatureDocumentId, originalContextBuffer.AsTextContainer());
+            foreach (var link in document.GetLinkedDocumentIds())
+            {
+                workspace.OpenDocument(link, originalContextBuffer.AsTextContainer());
+            }
+
+            // Start getting the compilation so the PartialSolution will be ready when the user starts typing in the window
+            await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+            _editorAdaptersFactoryService.SetDataBuffer(vsTextLines, projectionBuffer);
+
+            var result = new IntellisenseTextBoxViewModel[rolesCollections.Length];
+            for (var i = 0; i < rolesCollections.Length; i++)
+            {
+                var roleSet = _textEditorFactoryService.CreateTextViewRoleSet(rolesCollections[i]);
+
+                var vsTextView = _editorAdaptersFactoryService.CreateVsTextViewAdapter(_serviceProvider, roleSet);
+
+                vsTextView.Initialize(
+                    vsTextLines,
+                    IntPtr.Zero,
+                    (uint)TextViewInitFlags3.VIF_NO_HWND_SUPPORT,
+                    s_InitViews);
+
+                // Here we need ITextViewModelProvider handling the corresponding role.
```

Comment:
I admit it's not obvious to me how this comment is applying to the code below...?